File: basilisk-wiki/sandbox/acastillo/output_fields/tests_quantities/test_count_phase.py
#!/usr/bin/env python3
"""
Checker for `count_phase()`, shared by `test_count_phase.c` (`above = 1`) and
`test_count_phase_bubbles.c` (`above = 0`). Both use the same three spheres and
the same velocity field, so both expect the same numbers.

Row layout: `i t j volume bx by bz ux uy uz area` in 3D, dropping bz and uz in
2D; the dimension is inferred from the row width.

Closed form for sphere k, with u = U0 + Omega*z_hat x r:

  volume = 4/3*pi*R^3      b = c          area = 4*pi*R^2
  <u> = (U0x - Omega*cy, U0y + Omega*cx, U0z)

the average being exact because a sphere is symmetric about its centre and u is
linear. In 2D the same centres and radii give three disjoint discs, with
volume = pi*R^2 and area = 2*pi*R (a perimeter).

Regions are matched to spheres by volume, since `tag()` numbers them in
traversal order; the radii are distinct to make that unambiguous.

Volume and centre of mass converge fast; the interfacial area is the weak one,
being a sum over reconstructed VOF planes, hence its looser tolerance.

The constants below mirror the .c sources and must be kept in step by hand.

Usage:
  python3 test_count_phase.py [--droplets <file>] [--bubbles <file>]

Exits non-zero (and prints FAIL) when any region deviates by more than its
tolerance, or when the region count is wrong.
"""
import math
import sys

# Mirrors the .c sources
R_SPH = [0.18, 0.14, 0.10]
C_SPH = [(-0.25, -0.25, -0.25),
         (0.25, 0.25, -0.25),
         (0.20, -0.28, 0.28)]
U0 = (0.1, 0.2, 0.3)
OMEGA = 1.5

VOLUME_TOL = 0.01   # relative
AREA_TOL = 0.03     # relative
POS_TOL = 0.005     # absolute, in units of L0
VEL_TOL = 0.01      # absolute


def read_regions(filename):
  """Read the last time step's rows, skipping comment (`#`) and blank lines."""
  rows = []
  with open(filename) as f:
    for line in f:
      line = line.strip()
      if not line or line.startswith('#'):
        continue
      values = [float(v) for v in line.split()]
      if len(values) not in (9, 11):
        raise ValueError(f'unexpected row width {len(values)} '
                         f'(expected 9 in 2D or 11 in 3D)')
      rows.append(values)
  if not rows:
    return rows
  last = rows[-1][0]
  return [r for r in rows if r[0] == last]


def check_regions(filename, label):
  """Returns 0 on success, 1 on failure."""
  rows = read_regions(filename)
  if len(rows) != len(R_SPH):
    print(f'FAIL: {len(rows)} regions in {filename}, expected {len(R_SPH)}')
    return 1

  dim = 3 if len(rows[0]) == 11 else 2

  # Match regions to spheres by volume: tag() numbers them in traversal order.
  rows.sort(key=lambda r: r[3])
  order = sorted(range(len(R_SPH)), key=lambda k: R_SPH[k])

  worst = {'volume': 0., 'area': 0., 'b': 0., 'u': 0.}
  for row, k in zip(rows, order):
    volume, area = row[3], row[-1]
    b = row[4:4 + dim]
    u = row[4 + dim:4 + 2 * dim]
    r, c = R_SPH[k], C_SPH[k][:dim]

    if dim == 3:
      volume_exact = 4. / 3. * math.pi * r ** 3
      area_exact = 4. * math.pi * r ** 2
    else:
      volume_exact = math.pi * r ** 2
      area_exact = 2. * math.pi * r
    u_exact = (U0[0] - OMEGA * c[1], U0[1] + OMEGA * c[0], U0[2])[:dim]

    worst['volume'] = max(worst['volume'],
                          abs(volume - volume_exact) / volume_exact)
    worst['area'] = max(worst['area'], abs(area - area_exact) / area_exact)
    worst['b'] = max(worst['b'], max(abs(p - q) for p, q in zip(b, c)))
    worst['u'] = max(worst['u'], max(abs(p - q) for p, q in zip(u, u_exact)))

  print(f'Processed {len(rows)} regions in {dim}D ({label}). '
        f'Max relative error: volume {worst["volume"]:.3g}, '
        f'area {worst["area"]:.3g}. '
        f'Max absolute error: b {worst["b"]:.3g}, u {worst["u"]:.3g}')

  failures = []
  if worst['volume'] > VOLUME_TOL:
    failures.append('volume')
  if worst['area'] > AREA_TOL:
    failures.append('area')
  if worst['b'] > POS_TOL:
    failures.append('centre of mass')
  if worst['u'] > VEL_TOL:
    failures.append('velocity')

  if failures:
    print(f'FAIL: count_phase mismatch on {", ".join(failures)}.')
    return 1
  print('PASS: count_phase matches the sphere closed form.')
  return 0


CHECKS = {
  '--droplets': lambda f: check_regions(f, 'above = 1'),
  '--bubbles': lambda f: check_regions(f, 'above = 0'),
}

USAGE = ('Usage: python3 test_count_phase.py [--droplets <file>] '
         '[--bubbles <file>]')


def main():
  argv = sys.argv[1:]
  jobs = []  # (check_fn, filename), in the order given on the command line
  i = 0
  while i < len(argv):
    flag = argv[i]
    if flag not in CHECKS or i + 1 >= len(argv):
      print(USAGE)
      return 1
    jobs.append((CHECKS[flag], argv[i + 1]))
    i += 2

  if not jobs:
    print(USAGE)
    return 1

  status = 0
  for check, filename in jobs:
    print(f'--- {filename}')
    status |= check(filename)
  return status


if __name__ == '__main__':
  sys.exit(main())

File: basilisk-wiki/sandbox/acastillo/output_fields/tests_quantities/test_count_phase.c
/**
# Testing `count_phase()` on three spheres

Three well-separated spheres of distinct radii $R_k$ at centres
$\mathbf{c}_k$, in a solid-body rotation about $z$ superposed on a uniform
translation,

$$\mathbf{u} = \mathbf{U}_0 + \Omega\,\hat{z}\times\mathbf{r}$$

give every column of [count_phase.h](../count_phase.h) a closed form:
$V = \tfrac{4}{3}\pi R_k^3$, $\mathbf{b} = \mathbf{c}_k$,
$A = 4\pi R_k^2$, and -- since a sphere is symmetric about its centre, so the
volume-weighted average of a linear field is that field at the centre --
$\langle\mathbf{u}\rangle = \mathbf{U}_0 + \Omega\,\hat{z}\times\mathbf{c}_k$.
Distinct radii let the checker match regions to spheres by volume, `tag()`
numbering them in traversal order.

In 2D the same centres and radii describe three disjoint discs, with
$V = \pi R_k^2$ and $A = 2\pi R_k$ (a perimeter); the checker picks the right
closed form from the row width.

This is the `above = 1` (droplet) branch;
[test_count_phase_bubbles.c](test_count_phase_bubbles.c) covers `above = 0` on
the same geometry.

`count_phase()` writes `count_droplets.asc` itself, so nothing is computed
here: `test_count_phase.py` holds the closed form and the tolerances, and
mirrors the geometry below. */

#define LEVEL 5
#define MAXLEVEL 7

#include "navier-stokes/centered.h"
#include "two-phase.h"
#include "acastillo/output_fields/count_phase.h"

double R_sph[3] = {0.18, 0.14, 0.10};
coord c_sph[3] = {{-0.25, -0.25, -0.25},
                  { 0.25,  0.25, -0.25},
                  { 0.20, -0.28,  0.28}};

double U0_x = 0.1, U0_y = 0.2, U0_z = 0.3;
double Omega = 1.5;

// Level set of the union: positive inside any sphere.
double phi_spheres (double x, double y, double z) {
  double phi = -HUGE;
  for (int k = 0; k < 3; k++) {
    double d2 = sq(x - c_sph[k].x) + sq(y - c_sph[k].y);
#if dimension == 3
    d2 += sq(z - c_sph[k].z);
#endif
    double d = sqrt (d2);
    if (R_sph[k] - d > phi)
      phi = R_sph[k] - d;
  }
  return phi;
}

int main() {

  L0 = 1.0;
  X0 = Y0 = Z0 = -L0 / 2;
  N = 1 << LEVEL;
  init_grid (N);

#if TREE
  // Refine near the interfaces, where the VOF reconstruction sets the accuracy
  // of the volume and especially the area.
  double band = 4. * L0 / (1 << LEVEL);
  refine (fabs (phi_spheres (x, y, z)) < band && level < MAXLEVEL);
#endif

  fraction (f, phi_spheres (x, y, z));

  foreach() {
    u.x[] = U0_x - Omega * y;
    u.y[] = U0_y + Omega * x;
#if dimension == 3
    u.z[] = U0_z;
#endif
  }
  boundary ((scalar *){u});

  scalar m[];
  count_phase (m, 1e-6, 1, "count_droplets.asc", 0, 0.);

  /**
  The report goes to stderr, i.e. to the `log` diffed against
  `test_count_phase.ref`. */
  if (pid() == 0)
    system ("python3 ../test_count_phase.py "
            "--droplets count_droplets.asc 1>&2");
}
